fn/goroutine_manager.go

```diff
@@ -8,46 +8,55 @@ import (
 // GoroutineManager is used to launch goroutines until context expires or the
 // manager is stopped. The Stop method blocks until all started goroutines stop.
 type GoroutineManager struct {
-	wg     sync.WaitGroup
-	mu     sync.Mutex
-	ctx    context.Context
-	cancel func()
+	stopped sync.Once
+	wg      sync.WaitGroup
+	mu      sync.Mutex
+	quit    chan struct{}
 }
 
 // NewGoroutineManager constructs and returns a new instance of
 // GoroutineManager.
-func NewGoroutineManager(ctx context.Context) *GoroutineManager {
-	ctx, cancel := context.WithCancel(ctx)
-
+func NewGoroutineManager() *GoroutineManager {
 	return &GoroutineManager{
-		ctx:    ctx,
-		cancel: cancel,
+		quit: make(chan struct{}),
 	}
 }
 
 // Go tries to start a new goroutine and returns a boolean indicating its
-// success. It fails iff the goroutine manager is stopping or its context passed
-// to NewGoroutineManager has expired.
-func (g *GoroutineManager) Go(f func(ctx context.Context)) bool {
+// success. It returns true if the goroutine was successfully created and false
+// otherwise. A goroutine will fail to be created iff the goroutine manager is
+// stopping or the passed context has already expired. The passed call-back
+// function must exit if the passed context expires.
+func (g *GoroutineManager) Go(ctx context.Context,
+	f func(ctx context.Context)) bool {
+
+	// Derive a child context which will be canceled when either the passed
+	// context is canceled or the quit channel of the GoroutineManager is
+	// closed.
+	ctxc, cancel := ContextWithQuit(ctx, g.quit)
```

Comment:
`ContextWithQuit` starts another goroutine, which doubles the number of goroutines. (One worker + the one started by `ContextWithQuit`.)

Can context.WithCancel be used here instead? If parent context is also of type `cancelCtx`, no goroutine is created, judging from the [source](https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/context/context.go;l=478). Cancel functions can be stored in GoroutineManager; Stop() method would call all of them; each completed goroutine will remove its `cancel` function from the collection to prevent leaks.

Reply:
good idea - updated